codex/stores/repostore.nim

```diff
@@ -41,119 +45,277 @@ declareGauge(codex_repostore_bytes_reserved, "codex repostore bytes reserved")
-      return failure(newException(BlockNotFoundError, err.msg))
-    else:
-      return failure(err)
+  await self.metaDs.modifyT(metaKey,
+    proc (maybeCurrBlockMd: ?BlockMetadata): Future[?BlockMetadata] {.async.} =
+      if currBlockMd =? maybeCurrBlockMd:
+        BlockMetadata(
+          size: currBlockMd.size,
+          expiry: max(currBlockMd.expiry, minExpiry),
+          refCount: currBlockMd.refCount + plusRefCount - minusRefCount
+        ).some
+      else:
+        raise newException(BlockNotFoundError, "Metadata for block with cid " & $cid & " not found")
+  )
+
+proc storeBlock(self: RepoStore, blk: Block, minExpiry: SecondsSince1970): Future[?!StoreResult] {.async.} =
+  if blk.isEmpty:
+    return success(StoreResult(kind: AlreadyInStore))
+
+  without metaKey =? createBlockExpirationMetadataKey(blk.cid), err:
+    return failure(err)
+
+  without blkKey =? makePrefixKey(self.postFixLen, blk.cid), err:
+    return failure(err)
+
+  await self.metaDs.modifyTGetU(metaKey,
+    proc (maybeCurrMd: ?BlockMetadata): Future[(?BlockMetadata, StoreResult)] {.async.} =
+      var
+        md: BlockMetadata
+        res: StoreResult
+
+      if currMd =? maybeCurrMd:
+        if currMd.size == blk.data.len.NBytes:
+          md = BlockMetadata(size: currMd.size, expiry: max(currMd.expiry, minExpiry), refCount: currMd.refCount)
+          res = StoreResult(kind: AlreadyInStore)
+
+          # making sure that the block acutally is stored in the repoDs
+          without hasBlock =? await self.repoDs.has(blkKey), err:
+            raise err
+
+          if not hasBlock:
+            warn "Block metadata is present, but block is absent. Restoring block.", cid = blk.cid
+            if err =? (await self.repoDs.put(blkKey, blk.data)).errorOption:
+              raise err
+        else:
+          raise newException(CatchableError, "Repo already stores a block with the same cid but with a different size, cid: " & $blk.cid)
+      else:
+        md = BlockMetadata(size: blk.data.len.NBytes, expiry: minExpiry, refCount: 0)
+        res = StoreResult(kind: Stored, used: blk.data.len.NBytes)
+        if err =? (await self.repoDs.put(blkKey, blk.data)).errorOption:
+          raise err
 
-  return (Cid, MerkleProof).decodeCid(value)
+      (md.some, res)
+  )
+
+proc tryDeleteBlock(self: RepoStore, cid: Cid, expiryLimit = SecondsSince1970.high): Future[?!DeleteResult] {.async.} =
+  if cid.isEmpty:
+    return success(DeleteResult(kind: InUse))
+
+  without metaKey =? createBlockExpirationMetadataKey(cid), err:
+    return failure(err)
+
+  without blkKey =? makePrefixKey(self.postFixLen, cid), err:
+    return failure(err)
+
+  await self.metaDs.modifyTGetU(metaKey,
+    proc (maybeCurrMd: ?BlockMetadata): Future[(?BlockMetadata, DeleteResult)] {.async.} =
+      var
+        maybeMeta: ?BlockMetadata
+        res: DeleteResult
+
+      if currMd =? maybeCurrMd:
+        if currMd.refCount == 0 or currMd.expiry > expiryLimit:
```

Comment:
You're 100% right, there's a logic error here. It should be `currMd.expiry > expiryLimit` and the default value of `expiryLimit` should be `SecondsSince1970.low` instead of `SecondsSince1970.high`.

Thank you for catching this! I will add a test case for this.